### ISSUE TITLE
Model not converging: flat training/validation loss after 15 epochs

### ISSUE THREAD

**Issue by jeremymanning** [ContextLab/giblet-responses#32]
## Problem

After 15 epochs of training (3% of 500 total), both training and validation losses are essentially flat with minimal decrease:

- **Train Loss**: 275,824 - 276,118 (range: 294, or 0.11% variation)
  - Epoch 1: 275,959
  - Epoch 15: 275,825
  - **Net change: -134 (-0.05%)**

- **Validation Loss**: 204,916 - 210,534 (range: 5,618, or 2.7% variation)
  - Epoch 1: 207,225
  - Epoch 15: 206,622
  - **Net change: -603 (-0.29%)**

At current rate (~95.5 min/epoch), the full 500-epoch run will take ~32 days, but with such minimal loss reduction, this suggests the model is not learning effectively.

## Potential Causes to Investigate

1. **Learning rate issues**
   - Learning rate may be too low
   - Schedule may not be appropriate for this task
   - Warmup period configuration

2. **Gradient flow problems**
   - Gradients may be too small (vanishing gradients)
   - Initialization may be poor
   - Architecture may have gradient flow issues

3. **Data preprocessing issues**
   - fMRI data normalization/scaling
   - Video/audio feature normalization
   - Mismatch between data scales

4. **Loss function**
   - May not be providing sufficient gradient signal
   - Multiple loss components may be imbalanced

5. **Optimizer configuration**
   - Adam/AdamW parameters
   - Weight decay settings
   - Momentum/beta parameters

## Investigation Plan

- [ ] Check learning rate schedule configuration and actual LR values
- [ ] Analyze gradient magnitudes across model layers
- [ ] Review data preprocessing and normalization
- [ ] Verify loss function implementation
- [ ] Check optimizer configuration
- [ ] Create diagnostic scripts for gradient analysis
- [ ] Compare initialization strategies

## Training Configuration

- **Model**: Multimodal autoencoder (video + audio → fMRI voxels)
- **Dataset**: Sherlock fMRI (16 subjects, 50-minute stimulus)
- **Hardware**: 6 GPUs on tensor01
- **Epochs**: 500 total, currently at 15
- **Log file**: `logs/training_6gpu_20251104_223000.log`

**Comment by jeremymanning** [comment]
## Initial Diagnosis Complete

I've completed an initial investigation into the convergence issue. Here are the key findings:

### Major Findings

1. **Learning Rate Too Low During Warmup** ⚠️
   - Config uses 10-epoch warmup with LinearLR (0.1× → 1.0×)
   - At epoch 15, we're only at ~0.000015 LR (15% of target 0.0001)
   - Model needs higher LR earlier to make meaningful progress
   - **Recommendation**: Reduce warmup to 2-3 epochs

2. **Data Scale Mismatch** 🔴 **CRITICAL**
   - Video: normalized to [0, 1] → MSE ~ 0.01-1.0
   - Audio: discrete EnCodec codes [0-1024] → MSE ~ 1000-100000
   - Text: normalized embeddings [-1, 1] → MSE ~ 0.1-1.0
   - fMRI: **raw BOLD values [0-10000+]** → MSE ~ 1000000-100000000
   - **fMRI data is NOT normalized!** This is likely the primary issue.
   - fMRI loss dominates total loss by 6+ orders of magnitude
   - Model optimizes almost exclusively for fMRI, ignores reconstruction
   - **Recommendation**: Normalize fMRI data (z-score per subject) or reduce fMRI loss weight to ~0.001

3. **Loss Imbalance**
   - All loss weights set to 1.0 despite massive scale differences
   - fMRI loss contribution is millions of times larger than video/text losses
   - **Recommendation**: Either normalize all modalities OR rebalance weights

### Diagnostic Tools Created

Created `scripts/diagnose_training.py` to analyze:
- Data statistics (mean, std, range for each modality)
- Gradient flow through network
- Loss component magnitudes
- Learning rate schedule behavior
- Weight initialization

### Next Steps

1. Run diagnostic script on current checkpoint:
   ```bash
   python scripts/diagnose_training.py \
       --config configs/training/production_500epoch_config.yaml \
       --data-samples 100 \
       --output diagnostic_report.txt
   ```

2. Create test configs with fixes:
   - **Fix A**: Normalize fMRI data (add z-scoring to FMRIProcessor)
   - **Fix B**: Reduce warmup to 2-3 epochs
   - **Fix C**: Increase initial LR to 3e-4
   - **Fix D**: Rebalance loss weights (fMRI weight: 0.001)

3. Launch short test runs (10 epochs) to compare fixes

4. Choose best configuration and restart production run

### Files Created

- `scripts/diagnose_training.py` - Diagnostic script
- `notes/convergence_diagnosis_20251105.md` - Detailed findings
